src/Adapter/Cart/CommandHandler/AddCartRuleToCartHandler.php

```diff
@@ -59,13 +75,23 @@ public function handle(AddCartRuleToCartCommand $command)
         $cart = $this->getCart($command->getCartId());
         $cartRule = new CartRule($command->getCartRuleId()->getValue());
 
+        $this->contextStateManager
+            ->setCart($cart)
+            ->setCurrency(new Currency($cart->id_currency))
+            ->setLanguage(new Language($cart->id_lang))
+            ->setCustomer(new Customer($cart->id_customer));
+
         $errorMessage = $this->validateCartRule($cartRule, $cart);
 
         if ($errorMessage) {
+            $this->contextStateManager->restoreContext();
+
             throw new CartRuleValidityException($errorMessage);
         }
 
         if (!$cart->addCartRule($cartRule->id)) {
+            $this->contextStateManager->restoreContext();
+
             throw new CartException('Failed to add cart rule to cart.');
```

Comment:
Don't forget to call `$this->contextStateManager->restoreContext();` before the function returns

Reply:
Done